### ISSUE TITLE
[CIO] java.io.EOFException when parsing response to multi-part form data

### ISSUE THREAD

**Issue by serebit** [ktorio/ktor#1462]
**Ktor Version and Engine Used (client or server and name)**
1.2.6 and 1.3.0-beta-2, `ktor-client-cio`

**Describe the bug**
About 30% of the time, when sending a multi-part form data request to Discord's servers, the CIO client fails to parse the response sent by the server and ceases to function. The error thrown is `java.io.EOFException: Failed to parse HTTP response: unexpected EOF`, which causes a `java.io.IOException: Connection reset by peer`, which essentially softlocks the HttpClient. **This does not happen with `ktor-client-okhttp`, only CIO.** The trace is here: https://hastebin.com/cikecature.txt

**To Reproduce**
I can't provide an easy method of reproduction, as I'm not sure what part of the response is causing this behavior, and it could be specific to this scenario. If it helps, I can provide a packet capture from Wireshark.

**Expected behavior**
I expected the client to correctly parse the response.

**Comment by e5l** [comment]
Hi @serebit, thanks for the report. Could you provide the request example?

**Comment by serebit** [comment]
The request is a standard HTTP POST request with authorization headers, no URL parameters, and the following body construction:
```kotlin
MultiPartFormDataContent(formData {
    append("file", fileData, Headers.build {
        append(HttpHeaders.ContentDisposition,
            """form-data; name="file"; filename="$fileName"""")
    })
})
```
where `fileData` is a `ByteArray` read directly from an image file (a PNG with transparency).

**Comment by serebit** [comment]
If need be, I can provide a simplified single-block example of the call, although it won't be runnable due to how the Discord HTTP API works.

**Comment by e5l** [comment]
I failed to reproduce the issue. It looks like the issue is in the specific response parsing. It would be nice if you can provide some request/response captures for that case, but I continue investigation.

**Comment by dmitrievanthony** [comment]
Hi folks,

I spent some time and prepared the code that reproduces the problem, you can find it [here](https://github.com/dmitrievanthony/KPlayground/blob/%231462/src/main/kotlin/app.kt). It works fine with `OkHttp` and `Android` clients but fails with `CIO` with the following stack trace:

```
Exception in thread "main" java.io.EOFException: Failed to parse HTTP response: unexpected EOF
	at io.ktor.client.engine.cio.UtilsKt.readResponse(utils.kt:81)
	at io.ktor.client.engine.cio.Endpoint$makeDedicatedRequest$1$responseData$1.invokeSuspend(Endpoint.kt:102)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

I'll continue the investigation of the initial problem.

**Comment by serebit** [comment]
@dmitrievanthony Thanks for the reproduction!

**Comment by dmitrievanthony** [comment]
Hi, @serebit. I spent some time investigating the issue and found out that the root cause is TCP half-close behavior of CIO client. It's not supported by all servers (for example Jetty supports it, but Netty doesn't). Anyway, #1489 will fix it.

**Comment by serebit** [comment]
Is there a target release for this fix, @dmitrievanthony?

**Comment by dmitrievanthony** [comment]
Hi, @serebit. I've just merged the fix, so it will be included in `1.3.0`.

**Comment by serebit** [comment]
Yep, just updated to `1.3.0-rc` and hammered my test application with attachments, seems like that did it! Much appreciated.
